net: mdns: add capability of providing records in runtime

kderda opened the pull request:
Current implementation of mDNS responder does makes it mandatory to have all the records set at compile time. It is not suited well for applications that have to publish/unpublish or change records in runtime, e.g. data received from the network.

kderda committed:
net: mdns: add capability of providing records in runtime

Current implementation of mDNS responder does makes it mandatory to have
all the records set at compile time. It is not suited well for applications
that have to publish/unpublish or change records in runtime, e.g. data
received from the network.

Signed-off-by: Konrad Derda <konrad.derda@nordicsemi.no>

cfriedt commented:
> Current implementation of mDNS responder does makes it mandatory to have all the records set at compile time.

Technically, that's not exactly the case, it only requires the pointers to memory for DNS records to be set at compile time (and subsequently the number of records).

> It is not suited well for applications that have to publish/unpublish or change records in runtime, e.g. data received from the network.

The memory that each record points to can be runtime modifiable today.

In any case though, dns_sd is definitely ripe for improvement.

cfriedt commented on subsys/net/lib/dns/mdns_responder.c:
```diff
@@ -397,7 +401,21 @@ static void send_sd_response(struct net_context *ctx,
-	DNS_SD_FOREACH(record) {
+	DNS_SD_COUNT(&rec_num);
+
+	while (rec_num || ext_rec_num) {
```
Please use integer comparisons here instead of   treating the integers as bools

cfriedt reviewed (commented):
Looks good to me, with a couple of minor modifications.

Please fix the compliance issues and ensure the build is green.

If you have interest in this area, I would recommend checking out some of the items on the TODO list here https://github.com/zephyrproject-rtos/zephyr/issues/38846

cfriedt commented:
cc @pdgendt

rlubos commented on include/zephyr/net/mdns_responder.h:
```diff
@@ -0,0 +1,28 @@
+
+#include <stddef.h>
+
+#include "dns_sd.h"
```
`#include <zephyr/net/dns_sd.h>`

rlubos commented on include/zephyr/net/mdns_responder.h:
```diff
@@ -0,0 +1,28 @@
+ * @ref DNS_SD_REGISTER_SERVICE (if any) and then go over external records.
+ *
+ * @param records An array of mDNS records
+ * @param count The number of elemnts
```
```suggestion
 * @param count The number of elements
```

rlubos commented on include/zephyr/net/mdns_responder.h:
```diff
@@ -0,0 +1,28 @@
+ *
+ * @param records An array of mDNS records
+ * @param count The number of elemnts
+ * @return int 0 for OK; -EINVAL for invalid parameters.
```
`@return` doesn't need type

rlubos reviewed (commented):
Please check the compliance results, it reported some nits as well. Otherwise looks good.

konradderda commented:
Hello @cfriedt,
Thank you for your comments

> The memory that each record points to can be runtime modifiable today.

That was my first idea. However, it's not possible because the `text_size` field of `dns_sd_rec` is not a pointer so it would have a fixed value and cannot be modified: https://github.com/zephyrproject-rtos/zephyr/blob/e544bdb5e2ba254c4357c8835f26405b05c370d2/include/zephyr/net/dns_sd.h#L233

Moreover, such approach would be enough for a small number of records. With this approach the user can provide a pointer to a static array, memory slab, heap allocated memory, etc.

rlubos reviewed (commented):
Looking at the doc build output, it looks that we needed to escape square brackets in `struct dns_sd_rec` documentation (i. e. [\<](https://www.doxygen.nl/manual/commands.html#cmdlt) [\>](https://www.doxygen.nl/manual/commands.html#cmdgt))

konradderda commented:
@cfriedt @ssharks @tbursztyka, sorry for bothering you but it seems that this PR could be closed with an additional approval, could you please look at it? Thanks in advance.

ssharks reviewed (commented):
The only thing I observe: Is it clear to the users that the mdns_responder_set_ext_records knows it have to make the data stay at that location instead of providing a stack variable?

konradderda commented:
> The only thing I observe: Is it clear to the users that the mdns_responder_set_ext_records knows it have to make the data stay at that location instead of providing a stack variable?

@ssharks, Thank you for the question, I think that providing a stack variable is perfectly fine as long as the scope is prolonged - e.g. we can have an array of `dns_sd_rec` in the `main()` function with `while (true)` loop. Another example - delegating `k_work` to publish records only once for some time (e.g. for 5 minutes after initialization) where there is no point to occupy the RAM for whole time. These are not common use cases and of course one can use other solutions but I believe there is no need of enforcing protection from this (+ it feels like a common sense, like not returning pointers to stack variables).

jukkar commented on include/zephyr/net/mdns_responder.h:
```diff
@@ -0,0 +1,27 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
```
Please add multiple inclusion detection checks here, so
```
#ifndef ZEPHYR_INCLUDE_NET_MDNS_RESPONDER_H_
#define ZEPHYR_INCLUDE_NET_MDNS_RESPONDER_H_
```

konradderda replied:
Done

jukkar commented on include/zephyr/net/mdns_responder.h:
```diff
@@ -0,0 +1,27 @@
+ * @param count The number of elements
+ * @return 0 for OK; -EINVAL for invalid parameters.
+ */
+int mdns_responder_set_ext_records(struct dns_sd_rec records[], size_t count);
```
and add
```
#endif /* ZEPHYR_INCLUDE_NET_MDNS_RESPONDER_H_ */
```
here

konradderda replied:
Done

jukkar commented on include/zephyr/net/mdns_responder.h:
```diff
@@ -0,0 +1,27 @@
+ * mDNS responder will start with iteration over mDNS records registered using
+ * @ref DNS_SD_REGISTER_SERVICE (if any) and then go over external records.
+ *
+ * @param records An array of mDNS records
```
Please add documentation here about the fact that we are storing and using the user supplied pointer directly (not copying it), so the pointer must point to a memory that is valid for the time the records are being used. Typically this would mean that the records pointer should not be allocated from stack.

konradderda replied:
Done

jukkar reviewed (changes requested):
Some changes needed

jukkar reviewed (dismissed):
LGTM

cfriedt commented on subsys/net/lib/dns/mdns_responder.c:
```diff
@@ -691,4 +712,16 @@ static int mdns_responder_init(void)
 	return init_listener();
 }
 
+int mdns_responder_set_ext_records(struct dns_sd_rec records[], size_t count)
```
Please use a (const) pointer and size rather than the array and size.
```suggestion
int mdns_responder_set_ext_records(const struct dns_sd_rec *records, size_t count)
```

cfriedt commented on subsys/net/lib/dns/mdns_responder.c:
```diff
@@ -46,6 +46,8 @@ static struct net_context *ipv6;
 #endif
 
 static struct net_mgmt_event_callback mgmt_cb;
+static struct dns_sd_rec *external_records;
```
Unless the external records are modified in this compilation unit (which I assume is not the intent), this should probably be `const` (i.e. the data pointing to should not be modified).
```suggestion
static const struct dns_sd_rec *external_records;
```

cfriedt commented on include/zephyr/net/mdns_responder.h:
```diff
@@ -0,0 +1,34 @@
+ * @param count The number of elements
+ * @return 0 for OK; -EINVAL for invalid parameters.
+ */
+int mdns_responder_set_ext_records(struct dns_sd_rec records[], size_t count);
```
Please use a pointer-to-const and size instead of array and size.
```suggestion
int mdns_responder_set_ext_records(const struct dns_sd_rec *records, size_t count);
```

cfriedt reviewed (dismissed):
Some changes needed. Additionally, can you please add to coverage to the existing testsuite in a second commit?
See `tests/net/lib/dns_sd/src/main.c`

konradderda commented:
@jukkar I am sorry for the delay. There is no problem with addressing the comments regarding the code. However, I had tried to write a test case for my change and I found that with the current implementation it's quite hard to do. Unfortunately, at the moment I will not be able to provide a proper test implementation.

MaureenHelm commented:
@konradderda will you return to this before the 3.6 merge window closes this week?

jukkar commented:
ping @konradderda, are you able to continue with this? The needed changes look rather small and it would be nice to get this merged as this has been pending for a long time. It would be also nicer to get this in at the beginning of the 3.7 release cycle than at the end.

jukkar commented:
> The needed changes look rather small

Hmm, I missed the fact that unit tests would need changes too so the amount of work is bigger.

kderda committed:
tests: net: mdns: add test for external mDNS records

This commit adds a test for externally stored mDNS records for mDNS
responder.

Signed-off-by: Konrad Derda <konrad.derda@nordicsemi.no>

konradderda commented:
@jukkar @MaureenHelm I had finally got some time and prepared a test for this change with a foundations for implementation of new test cases in the futute. Also, I have applied last remarks from @cfriedt .

jukkar commented on tests/net/lib/mdns_responder/src/main.c:
```diff
@@ -0,0 +1,445 @@
+	net_ipv6_nbr_add(iface1, &sender_ll_addr, net_if_get_link_addr(iface1), false,
+			 NET_IPV6_NBR_STATE_STATIC);
+
+	if (!ifaddr) {
+		zassert_not_null(ifaddr, "Failed to add LL-addr");
+	} else {
+		/* we need to set the addresses preferred */
+		ifaddr->addr_state = NET_ADDR_PREFERRED;
+	}
```
We can simplify this to:
```
zassert_not_null(ifaddr, "Failed to add LL-addr");

/* we need to set the addresses preferred */
ifaddr->addr_state = NET_ADDR_PREFERRED;
```
because if the zassert fails, then the program stops so there is no need to have the if() else() branches.

konradderda replied:
Done

jukkar commented on subsys/net/lib/dns/mdns_responder.c:
```diff
@@ -423,7 +428,21 @@ static void send_sd_response(struct net_context *ctx,
+		 */
+		if (rec_num > 0) {
+			DNS_SD_GET(rec_num - 1, &record);
+			rec_num -= 1;
```
That is perfectly valid code but usually -1 is done via `rec_num--;`

konradderda replied:
Done

jukkar commented on subsys/net/lib/dns/mdns_responder.c:
```diff
@@ -423,7 +428,21 @@ static void send_sd_response(struct net_context *ctx,
+			rec_num -= 1;
+		} else {
+			record = &external_records[ext_rec_num - 1];
+			ext_rec_num -= 1;
```
and `ext_rec_num--;`

jukkar commented on subsys/net/lib/dns/mdns_responder.c:
```diff
@@ -423,7 +428,21 @@ static void send_sd_response(struct net_context *ctx,
-	DNS_SD_FOREACH(record) {
+	DNS_SD_COUNT(&rec_num);
+
+	while (rec_num > 0 || ext_rec_num > 0) {
```
So this means that if there is different count of records, the minimum one wins and the rest of the other records are skipped. Just wanted to verify that this is the intention?
